### ISSUE TITLE
Add Volume-Under-Surface (VUS) metrics

### ISSUE THREAD

**Issue by CodeLionX** [HPI-Information-Systems/TimeEval#30]
Four new metrics have recently [1] been introduced:
- Range-AUC-ROC
- Range-AUC-PR
- VUS-ROC
- VUS-PR

Add those metrics to TimeEval. The Implementation can be found here: https://github.com/TheDatumOrg/VUS

> [1]: John Paparrizos, Paul Boniol, Themis Palpanas, Aaron Elmore, and Michael J. Franklin. Volume Under the Surface: new accuracy measures for abnormal subsequences detection in time series. PVLDB, 15(X): 2774 - 2787, 2022. doi:10.14778/3551793.3551830

**Comment by B-Seif** [comment]
Hi,
Can we have the results of your first experiment (WenigEtAl2022TimeEval) according to these new metrics?

How we can integrate them without changing the version of TimeEval ? 

we have a cluster that runs under timeval 1.4.2 and we are afraid of breaking everything if we switch to another version.

**Comment by CodeLionX** [comment]
Hi @B-Seif,

We recalculated the scores of our benchmark experiments in November last year. However, we did not use the compatibility mode (`compatibility_mode=True`, see [docu](https://timeeval.readthedocs.io/en/latest/api/timeeval.metrics.html#timeeval-metrics-rangeprauc)). You can download the results from [here](https://nextcloud.hpi.de/s/epJAZParcnjMkPb) **until the end of March '23**. Password: `%NDzTVmng$$7wHm!rZoYSw^9Rvv:@2px`.

You can just use a second conda/venv/pyenv environment with a newer version of TimeEval. From this env, you can call the [`scripts/recalculate_metrics.py`](https://github.com/HPI-Information-Systems/TimeEval/blob/main/scripts/recalculate_metrics.py) script to recompute the metrics from an already existing TimeEval run (experiment results folder). **Please create a backup of your folder** before using this script because it modifies the folder in-place and is not tested as well as the rest of the system.

**Comment by B-Seif** [comment]
Hi @CodeLionX ,
Thank you very much for sharing the results file.

Is the code below compatible with the new version of TimeEval? and what do I need to modify to add all the evaluation measures you have implemented (AUC-ROC, RANGE-ROC-AUC, RANGE-PRAUC, F1,PR,RE, Precision@K,...)

```python
#!/usr/bin/env python3
import timeeval
assert timeeval.__version__ == "1.2.4", "TimeEval version not supported. This script is for TimeEval version 1.2.4!"
from pathlib import Path
from timeeval import TimeEval, MultiDatasetManager, Metric, Algorithm, TrainingType, InputDimensionality, ResourceConstraints
from timeeval.adapters import DockerAdapter
from timeeval.params import FixedParameters
from timeeval.resource_constraints import GB
from timeeval.params import FullParameterGrid
from timeeval.utils.window import ReverseWindowing
from timeeval import TimeEval, RemoteConfiguration, ResourceConstraints
import numpy as np
from durations import Duration
def main():
    # cluster setup
    cluster_config = RemoteConfiguration(
    scheduler_host="192.168.100.4",
    worker_hosts=[ "192.168.100.1","192.168.100.2","192.168.100.3","192.168.100.4"]
  
     )
    dm = MultiDatasetManager([Path("/home/ad/timeeval-datasets/")])
    datasets = dm.select() 
    algorithms = [
        Algorithm(
            name= "detector",
            main=DockerAdapter(image_name="registry.gitlab.hpi.de/akita/i/detector",tag="0.1.2", skip_pull=True),
            data_as_file=True,
            param_config=FullParameterGrid({
                "window_size" : [20,50],}),
            training_type=TrainingType.UNSUPERVISED,
            input_dimensionality=InputDimensionality.MULTIVARIATE
        ) ]
    repetitions = 1
    rcs = ResourceConstraints(
    task_memory_limit = 3 * GB,
    task_cpu_limit = 1.0,
    tasks_per_host=10,
    execute_timeout=Duration("2 hours")
)
    

    timeeval = TimeEval(
        dm, 
        datasets, 
        algorithms,
        repetitions=repetitions, 
        metrics=[Metric.ROC_AUC,Metric.PR_AUC,Metric.RANGE_PR_AUC,.....], # from where can I add other eval measures 
        remote_config=cluster_config,
        resource_constraints=rcs,
        distributed=True
    )
    timeeval.run()
    results = timeeval.get_results(aggregated=True)
    print(results)

if __name__ == "__main__":
    main()
```

**Comment by CodeLionX** [comment]
Most of your code above should be compatible with the newest version of TimeEval (as of now: 1.2.10). However, the "old" default metrics can now be found at `timeeval.DefaultMetrics`. All other metrics must be imported from `timeeval.metrics`:

```python
from timeeval import DefaultMetrics
from timeeval.metrics import RangePrAUC, [...]

metrics = [
    DefaultMetrics.ROC_AUC,
    DefaultMetrics.PR_AUC,
    DefaultMetrics.RANGE_PR_AUC,
    RangePrAUC(buffer_size=100, compatibility_mode=True),
    [...]
]
```

You can find all available metrics in the docs: https://timeeval.readthedocs.io/en/latest/api/timeeval.metrics.html

---

Don't be surprised by an additional status `Status.OOM = 3` in the results, though. Since TimeEval v1.2.10, the `DockerAdapter` now handles memory errors similar to timeout errors: #61

**Comment by B-Seif** [comment]
Thank you for this information.

I really want to migrate to the new version of TimeEval but I'm afraid my cluster can't restart properly.
Can I just create a new environment (conda) in each machine and install TimeEval ? or I really have to delete everything and put the new TimeEval in the each machine (without going through the environment).

**Comment by B-Seif** [comment]
Hi @CodeLionX 

Please how can i use the range_roc_vus metric like the default one as in the code below : 

```python
from timeeval import DefaultMetrics
from timeeval.metrics import RangePrAUC,RangeRocAUC,RangeRocVUS

roc = DefaultMetrics.ROC_AUC(df_data["is_anomaly"], scores)
pr  = DefaultMetrics.PR_AUC(df_data["is_anomaly"], scores)
rpr = DefaultMetrics.RANGE_PR_AUC(df_data["is_anomaly"], scores)

rocV =RangeRocVUS(df_data["is_anomaly"], scores,max_buffer_size=100,compatibility_mode =False,max_samples=250)

print("roc :",roc,"\npr:",pr,"\nrpr:",rpr,"\nroc_vus:",rocV)

```

**Comment by CodeLionX** [comment]
Ah, yes. The documentation is a bit brief in this area. We should add a small example on how to use the metrics outside of TimeEval.

### tl;dr

```python
import numpy as np
from timeeval import DefaultMetrics
from timeeval.metrics import RangePrAUC

rng = np.random.default_rng(42)
y_true = rng.random(100) > 0.5
y_score = rng.random(100)
metrics = []

# default metrics are already instantiated objects:
m = DefaultMetrics.ROC_AUC
metrics.append(m)

# all metrics (in general) are classes that need to be instantiated with their parametrization:
m = RangePrAUC(buffer_size=100)
metrics.append(m)

# compute the metrics
for m in metrics:
    metric_value = m(y_true, y_score)
    print(f"{m.name} = {metric_value}")
```

**Comment by B-Seif** [comment]
Thank you for your return, it is clear.

I have a question about the results ([file here](https://nextcloud.hpi.de/s/epJAZParcnjMkPb/authenticate/showShare)).



Indeed, I noticed that for some algorithms there are NaN values.

Does this mean they don't meet the constraints (1 core, 3 GB and 2 hours)?
 
See the attached image please (results for PR_AUC measure for example ).


![image](https://user-images.githubusercontent.com/21296530/221543072-9845d3fc-31fb-40a6-9fde-5a4ca4bb4b30.png)

**Comment by CodeLionX** [comment]
> Thank you for your return, it is clear.
> 
> I have a question about the results ([file here](https://nextcloud.hpi.de/s/epJAZParcnjMkPb/authenticate/showShare)).
> 
> Indeed, I noticed that for some algorithms there are NaN values.
> 
> Does this mean they don't meet the constraints (1 core, 3 GB and 2 hours)?
> 
> See the attached image please (results for PR_AUC measure for example ).
> 
> ![image](https://user-images.githubusercontent.com/21296530/221543072-9845d3fc-31fb-40a6-9fde-5a4ca4bb4b30.png)

Indeed, NaNs mean that we don't have the algorithm's results to compute the metric. There could be multiple reasons for that:

- Algorithm failed
- Timeout
- Memory Error

You can inspect the `status` and `error_message` columns for more details. But the errors should be consistent with the original results.
